rust/main/chains/hyperlane-sealevel/src/provider.rs

```diff
@@ -2,37 +2,110 @@ use std::sync::Arc;
 use hyperlane_core::{
     BlockInfo, ChainCommunicationError, ChainInfo, ChainResult, HyperlaneChain, HyperlaneDomain,
-    HyperlaneProvider, HyperlaneProviderError, TxnInfo, TxnReceiptInfo, H256, H512, U256,
+    HyperlaneProvider, HyperlaneProviderError, NativeToken, TxnInfo, TxnReceiptInfo, H256, H512,
+    U256,
 };
 
 use crate::error::HyperlaneSealevelError;
 use crate::utils::{decode_h256, decode_h512, decode_pubkey};
 use crate::{ConnectionConf, SealevelRpcClient};
 
+/// Exponent value for atto units (10^-18).
+const ATTO_EXPONENT: u32 = 18;
+
 /// A wrapper around a Sealevel provider to get generic blockchain information.
 #[derive(Debug)]
 pub struct SealevelProvider {
     domain: HyperlaneDomain,
     rpc_client: Arc<SealevelRpcClient>,
+    native_token: NativeToken,
 }
 
 impl SealevelProvider {
     /// Create a new Sealevel provider.
     pub fn new(domain: HyperlaneDomain, conf: &ConnectionConf) -> Self {
         // Set the `processed` commitment at rpc level
         let rpc_client = Arc::new(SealevelRpcClient::new(conf.url.to_string()));
+        let native_token = conf.native_token.clone();
 
-        SealevelProvider { domain, rpc_client }
+        SealevelProvider {
+            domain,
+            rpc_client,
+            native_token,
+        }
     }
 
     /// Get an rpc client
     pub fn rpc(&self) -> &SealevelRpcClient {
         &self.rpc_client
     }
+
+    fn validate_transaction(hash: &H512, txn: &UiTransaction) -> ChainResult<()> {
+        let received_signature = txn
+            .signatures
+            .first()
+            .ok_or(HyperlaneSealevelError::UnsignedTransaction(*hash))?;
+        let received_hash = decode_h512(received_signature)?;
+
+        if &received_hash != hash {
+            Err(Into::<ChainCommunicationError>::into(
+                HyperlaneSealevelError::IncorrectTransaction(
+                    Box::new(*hash),
+                    Box::new(received_hash),
+                ),
+            ))?;
+        }
+        Ok(())
+    }
+
+    fn sender(hash: &H512, txn: &UiTransaction) -> ChainResult<H256> {
+        let message = match &txn.message {
+            UiMessage::Parsed(m) => m,
+            m => Err(Into::<ChainCommunicationError>::into(
+                HyperlaneSealevelError::UnsupportedMessageEncoding(m.clone()),
+            ))?,
+        };
+
+        let signer = message
+            .account_keys
+            .first()
+            .ok_or(HyperlaneSealevelError::UnsignedTransaction(*hash))?;
+        let pubkey = decode_pubkey(&signer.pubkey)?;
+        let sender = H256::from_slice(&pubkey.to_bytes());
+        Ok(sender)
+    }
+
+    fn gas(meta: &UiTransactionStatusMeta) -> ChainResult<U256> {
+        let OptionSerializer::Some(gas) = meta.compute_units_consumed else {
+            Err(HyperlaneSealevelError::EmptyComputeUnitsConsumed)?
+        };
+
+        Ok(U256::from(gas))
+    }
+
+    fn fee(&self, meta: &UiTransactionStatusMeta) -> ChainResult<U256> {
```

Comment:
We need it since the compute unit (gas unit) costs less than 1 lamport on Solana and we need to express the gas price in an integer number. So, we convert all fees into units of 10^-18 (same for all the chains) so that gas price is meaningful.

Reply:
I added a comment into the source code.